aspects/schema.go

```diff
@@ -0,0 +1,358 @@
+	if rawEntries, ok := constraints["schema"]; ok {
+		var entries map[string]json.RawMessage
+		if err := json.Unmarshal(rawEntries, &entries); err != nil {
+			return fmt.Errorf(`cannot parse map's "schema" constraint: %v`, err)
+		}
+
+		v.entrySchemas = make(map[string]Schema, len(entries))
+		for key, value := range entries {
+			entrySchema, err := v.topSchema.parse(value)
+			if err != nil {
+				return err
+			}
+
+			v.entrySchemas[key] = entrySchema
+		}
+
+		// "required" can be a list of keys or many lists of alternative combinations
+		if rawRequired, ok := constraints["required"]; ok {
+			var requiredCombs [][]string
+			if err := json.Unmarshal(rawRequired, &requiredCombs); err != nil {
+				var typeErr *json.UnmarshalTypeError
+				if !errors.As(err, &typeErr) {
+					return fmt.Errorf(`cannot parse map's "required" constraint: %v`, err)
+				}
+
+				var required []string
+				if err := json.Unmarshal(rawRequired, &required); err != nil {
+					return fmt.Errorf(`cannot parse map's "required" constraint: %v`, err)
+				}
+
+				v.requiredCombs = [][]string{required}
+			} else {
+				v.requiredCombs = requiredCombs
+			}
+		}
+
+		return nil
+	}
+
+	// map can not specify "schemas" and constrain the type of keys and values instead
+	rawKeyDef, ok := constraints["keys"]
+	if ok {
+		if v.keySchema, err = v.parseMapKeyType(rawKeyDef); err != nil {
+			return fmt.Errorf(`cannot parse "keys" constraint: %w`, err)
+		}
+	}
+
+	rawValuesDef, ok := constraints["values"]
+	if ok {
+		v.valueSchema, err = v.topSchema.parse(rawValuesDef)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// checkExclusiveConstraints checks if the map contains mutually exclusive constraints.
+func checkExclusiveConstraints(obj map[string]json.RawMessage) error {
+	has := func(k string) bool {
+		_, ok := obj[k]
+		return ok
+	}
+
+	if has("required") && !has("schema") {
+		return fmt.Errorf(`cannot use "required" without "schema" constraint`)
+	}
+	if has("schema") && has("keys") {
+		return fmt.Errorf(`cannot use "schema" and "keys" constraints simultaneously`)
+	}
+	if has("schema") && has("values") {
+		return fmt.Errorf(`cannot use "schema" and "values" constraints simultaneously`)
+	}
+
+	return nil
+}
+
+func (v *mapSchema) parseMapKeyType(raw json.RawMessage) (Schema, error) {
+	var typ string
+	if err := json.Unmarshal(raw, &typ); err != nil {
+		var typeErr *json.UnmarshalTypeError
+		if !errors.As(err, &typeErr) {
+			return nil, err
+		}
+
+		var schemaDef map[string]json.RawMessage
+		if err := json.Unmarshal(raw, &schemaDef); err != nil {
+			return nil, err
+		}
+
+		if rawType, ok := schemaDef["type"]; ok {
+			if err := json.Unmarshal(rawType, &typ); err != nil {
+				return nil, err
+			}
+
+			if typ != "string" {
+				return nil, fmt.Errorf(`must be based on string but got %q`, typ)
+			}
+		}
+
+		schema := &stringSchema{}
+		if err := schema.parseConstraints(schemaDef); err != nil {
+			return nil, err
```

Comment:
we cannot reach here because stringSchema.parseConstraints is always err = nil? should we leave a TODO?

Reply:
Yes, exactly, since there aren't string constraints yet this can't fail. The next PR will add constraints and the failing case will cover this